keps/0021-kudo-upgrade.md

```diff
@@ -0,0 +1,271 @@
+
+## Open Questions
+- Lowest supported K8s version
+- Do we want to support downgrades? 
+- Split versions between KUDO manager and KUDO CLI?
+
+## Proposal
+
+### KUDO Prerequisites
+Expected update frequency: Low  
+Versioned: No, but closely tied to KUDO manager
+
+The KUDO manager has a set of prerequisites that need to be present for the manager to run successfully. They are
+the least likely to change, but probably the most specific. If we make changes here, we need to implement custom
+migration code.
+
+- Current prereqs
+  - Namespace
+  - Service Account
+  - Role Bindings
+  - Secret
+
+#### Proposal for update process 
+Integrated into `kudo init --upgrade`
+
+Write specific migration code that targets a KUDO manager version range and executes manual migration steps. 
+
+Each migration should have a validate-step that checks if the migration is possible.
+
+**Alternative update process**
+- Can we just delete them all and reinstall them? Probably not
+
+### KUDO Manager
+Expected update frequency: High  
+Versioned: Yes
+
+The KUDO Manager is defined by an image version in a deployment set. To update, the deployment must be updated. The 
+manager is closely tied to the CRDs, but not to the CLI. When CRDs are updated, the Manager will most likely also
+need to be updated. 
+
+#### Proposal for update process
+Integrated into `kudo init --upgrade`
+
+- Use semantic versioning for the manager binary
+  - ToBeDiscussed: Use different versioning for Manager and CLI?
+- The manager needs to be shut down before updating the CRDs
+  - As soon as the new CRD is installed and marked as "stored"
+  - Without WebHook Conversion, even if the new CRD version is "served", the old content is returned without any conversion
+- Question: Do we need to ensure that the manager is not doing meaningful work at the moment, or can we just update the deployment?
+
+### CRDs
+Expected update frequency: Medium  
+Versioned: Yes, with a CRD-Version
+
+The CRDs are used to store installed operators and running instances. New features will regularly require us to add new
+fields. 
+- Existing CRs need to be migrated to new versions
+- K8s CRD support
+  - Which K8s version do we want/need to support?
+    - MultiVersion is supported since 1.11 (manual conversion)
+    - WebHook conversion since 1.16
+  - WebHook conversion would allow us to transparently switch to a new CRD version without manually migrating all existing CRs
+- CRD versioning
+  - Currently we have one global version for all CRDs
+  - Alternatively, we could have a one version per CRD
+  - Single version should be ok, in the worst case we have empty migrations
+  - One version for each CRD would require complex compatibility matrix
+
+#### Proposal for update process
+Integrated into `kudo init --upgrade`
+
+To support older K8s versions, no WebHook conversion is used. We only serve a single CRD-Version and migrate existing CRs in the update process
```

Comment:
I think we're fine in terms of kube version here `Webhook conversion is available as beta since 1.15, and as alpha since Kubernetes 1.13.`

Reply:
Which means, the operating cluster needs to have webhook conversion explicitly enabled for KUDO to work. I'm not sure if we want to make this a requirement for KUDO, tbh.